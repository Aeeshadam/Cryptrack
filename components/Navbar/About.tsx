import React from "react";
import { useSelector } from "react-redux";
import { AppState } from "@/store/store";
import { Stack, Typography } from "@mui/material";
import Box from "@mui/material/Box";

const About = () => {
  const { coin } = useSelector((state: AppState) => state.coinDetails);
  const about = coin?.description?.en;

  return (
    <Box bgcolor="secondary.main" marginY="2rem" padding="2rem">
      <Typography variant="h3">About</Typography>
      <Typography variant="body1" color="grey">
        {about}
      </Typography>
      <Typography marginTop="2rem" variant="subtitle1" fontWeight="600">
        Markets
      </Typography>
      <Stack
        direction="row"
        gap="1rem"
        marginTop="1rem"
        justifyContent="space-between"
        flexWrap="wrap"
      >
        {coin?.tickers.slice(0, 10).map((ticker: any) => (
          <Typography
            color="primary.main"
            fontWeight={500}
            key={ticker.market?.name}
            variant="body2"
          >
            {ticker.market?.name}
          </Typography>
        ))}
      </Stack>
    </Box>
  );
};
export default About;
