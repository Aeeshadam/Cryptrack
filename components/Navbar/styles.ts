import { styled } from "@mui/material/styles";
import InputBase from "@mui/material/InputBase";
import AppBar from "@mui/material/AppBar";
import Toolbar from "@mui/material/Toolbar";

export const StyledAppBar = styled(AppBar)(({ theme }) => ({
  backgroundColor: "transparent",
  backgroundImage: "none",
  boxShadow: "none",
}));

export const StyledToolbar = styled(Toolbar)(({ theme }) => ({
  display: "flex",
  justifyContent: "space-between",
  alignItems: "center",
  paddingLeft: 0,
  paddingRight: 0,
}));

export const Search = styled("div")(({ theme }) => ({
  position: "relative",
  borderColor: theme.palette.primary.main,
  color: theme.palette.common.white,
  borderWidth: 0.5,
  width: "200px",
  borderStyle: "solid",
  margin: theme.spacing(2),
  [theme.breakpoints.up("sm")]: {
    display: "block",
  },
  [theme.breakpoints.down("sm")]: {
    display: "none",
  },
}));

export const SearchIconWrapper = styled("div")(({ theme }) => ({
  padding: theme.spacing(0, 1),
  height: "100%",
  position: "absolute",
  pointerEvents: "none",
  display: "flex",
  alignItems: "center",
  justifyContent: "center",
}));

export const MobileSearchIconWrapper = styled("div")(({ theme }) => ({
  display: "flex",
  alignItems: "center",
  justifyContent: "flex-end",
  border: "1px solid ",
  borderColor: theme.palette.primary.main,
  width: "100%",
  cursor: "pointer",
  padding: theme.spacing(0.8),
}));

export const StyledInputBase = styled(InputBase)(({ theme }) => ({
  color: "inherit",
  "& .MuiInputBase-input": {
    padding: theme.spacing(1, 1, 1, 0),
    paddingLeft: `calc(1em + ${theme.spacing(3)})`,
    transition: theme.transitions.create("width"),
    width: "100%",
    [theme.breakpoints.up("md")]: {
      width: "20ch",
    },
  },
}));
