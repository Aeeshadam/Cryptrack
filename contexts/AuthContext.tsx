"use client";
import React, { useEffect } from "react";
import {
  createUserWithEmailAndPassword,
  signInWithEmailAndPassword,
  signOut,
  onAuthStateChanged,
  User,
  updateProfile,
} from "firebase/auth";
import { doc, setDoc, getDoc } from "firebase/firestore";
import db, { auth } from "@/firebase/firebase";
import { createContext, useContext, useState, ReactNode } from "react";
import LoadingSpinner from "@/components/LoadingSpinner";

interface AuthContextProps {
  user: User | null;

  signUp: (email: string, password: string, name: string) => Promise<boolean>;
  signIn: (email: string, password: string) => Promise<boolean>;
  loading: boolean;
  logOut: () => Promise<boolean>;
  displayName: string;
  setError: (error: string) => void;
  // setDisplayName: (name: string | null) => void;
  error: string;
}
const AuthContext = createContext<AuthContextProps | undefined>(undefined);

export const AuthProvider = ({ children }: { children: ReactNode }) => {
  const [user, setUser] = useState<User | null>(null);
  const [loading, setLoading] = useState(true);
  const [displayName, setDisplayName] = useState<string>("");
  const [error, setError] = useState<string>("");

  useEffect(() => {
    const unsubscribe = onAuthStateChanged(auth, async (user) => {
      if (user) {
        setUser(user);
        const docRef = doc(db, "users", user.uid);
        const docSnap = await getDoc(docRef);
        if (docSnap.exists()) {
          setDisplayName(docSnap.data().name);
        }
      } else {
        setUser(null);
        setDisplayName("");
      }
      setLoading(false);
    });
    return unsubscribe;
  }, []);

  const signUp = async (email: string, password: string, name: string) => {
    setLoading(true);
    setError("");
    try {
      const userCredential = await createUserWithEmailAndPassword(
        auth,
        email,
        password
      );
      const user = userCredential.user;
      await updateProfile(user, { displayName: name });
      await setDoc(doc(db, "users", user.uid), {
        name,
        email,
      });
      setDisplayName(name);
      setUser(user);
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
      setLoading(false);
    }
  };

  const signIn = async (email: string, password: string) => {
    setLoading(true);
    setError("");
    try {
      const userCredential = await signInWithEmailAndPassword(
        auth,
        email,
        password
      );
      const user = userCredential.user;

      const docRef = doc(db, "users", user.uid);
      const docSnap = await getDoc(docRef);

      if (docSnap.exists()) {
        setDisplayName(docSnap.data().name);
      }
      setUser(user);
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
      setLoading(false);
    }
  };

  const logOut = async () => {
    setError("");
    try {
      await signOut(auth);
      setUser(null);
      setDisplayName("");
      return true;
    } catch (error) {
      setError((error as Error).message);
      return false;
    } finally {
    }
  };

  return (
    <AuthContext.Provider
      value={{
        user,
        logOut,
        setError,
        loading,
        error,
        signUp,
        signIn,
        displayName,
      }}
    >
      {children}
    </AuthContext.Provider>
  );
};

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
