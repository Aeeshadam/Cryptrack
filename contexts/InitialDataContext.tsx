"use client";

import {
  createContext,
  useContext,
  useState,
  useEffect,
  ReactNode,
} from "react";
import { InitialDataContextProps, CoinProps } from "@/types";
import { useDispatch } from "react-redux";
import { AppDispatch } from "@/store/store";

const InitialDataContext = createContext<InitialDataContextProps | undefined>(
  undefined
);

export const InitialDataProvider = ({ children }: { children: ReactNode }) => {
  const [initialData, setInitialData] = useState([] as CoinProps[]);
  const dispatch = useDispatch<AppDispatch>();
  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(
          `${process.env.NEXT_PUBLIC_BASE_URL}/api/coins`
        );
        if (!response.ok) {
          throw new Error("Failed to fetch coins");
        }
        const data: CoinProps[] = await response.json();
        setInitialData(data);
      } catch (error) {
        console.error("Failed to fetch initial data:", error);
      }
    };

    // fetchData();
  }, [dispatch]);

  return (
    <InitialDataContext.Provider value={{ initialData, setInitialData }}>
      {children}
    </InitialDataContext.Provider>
  );
};

export const useInitialData = () => {
  const context = useContext(InitialDataContext);

  if (!context) {
    throw new Error("useInitialData must be used within a InitialDataProvider");
  }

  return context;
};

export default InitialDataContext;
