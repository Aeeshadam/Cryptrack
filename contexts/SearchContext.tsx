"use client";

import { createContext, useContext, useState, ReactNode } from "react";
import { SearchContextProps } from "@/types";
import { useSelector } from "react-redux";
import { AppState } from "@/store/store";

const SearchContext = createContext<SearchContextProps | undefined>(undefined);

export const SearchProvider = ({ children }: { children: ReactNode }) => {
  const [openSearch, setOpenSearch] = useState(false);
  const [searchQuery, setSearchQuery] = useState("");
  const coins = useSelector((state: AppState) => state.crypto.coins);

  const filteredCoins = searchQuery
    ? coins?.filter((coin) =>
        coin.name.toLowerCase().includes(searchQuery.toLowerCase())
      )
    : coins?.slice(0, 10);

  const handleOpenSearch = () => {
    setOpenSearch(true);
  };
  return (
    <SearchContext.Provider
      value={{
        searchQuery,
        setSearchQuery,
        filteredCoins,
        openSearch,
        setOpenSearch,
        handleOpenSearch,
      }}
    >
      {children}
    </SearchContext.Provider>
  );
};

export const useSearch = () => {
  const context = useContext(SearchContext);
  if (context === undefined) {
    throw new Error("useSearch must be used within a SearchProvider");
  }
  return context;
};
