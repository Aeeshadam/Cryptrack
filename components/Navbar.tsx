"use client";
import React from "react";
import Box from "@mui/material/Box";
import Logo from "./Logo";
import SearchComponent from "./SearchComponent";
import { StyledAppBar, StyledToolbar } from "../styles/NavbarStyles";
import { StyledButton } from "./StyledButton";

export default function Navbar() {
  return (
    <>
      <Box sx={{ flexGrow: 1 }}>
        <StyledAppBar position="static">
          <StyledToolbar style={{ padding: 0 }}>
            <Logo />
            <SearchComponent />
            <StyledButton size="small">Get Started</StyledButton>
          </StyledToolbar>
        </StyledAppBar>
      </Box>
    </>
  );
}
