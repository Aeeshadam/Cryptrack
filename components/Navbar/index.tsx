"use client";
import Box from "@mui/material/Box";
import Logo from "../Logo";
import SearchComponent from "../SearchComponent";
import SearchModal from "../SearchModal";
import { StyledAppBar, StyledToolbar } from "./styles";
import { StyledButton } from "../StyledButton";
import { useState } from "react";

export default function Navbar() {
  const [openSearch, setOpenSearch] = useState(false);
  return (
    <>
      {openSearch && (
        <SearchModal setOpenSearch={setOpenSearch} open={openSearch} />
      )}
      <Box sx={{ flexGrow: 1 }}>
        <StyledAppBar position="static">
          <StyledToolbar style={{ padding: 0 }}>
            <Logo />
            <SearchComponent setOpenSearch={setOpenSearch} />
            <StyledButton size="small">Get Started</StyledButton>
          </StyledToolbar>
        </StyledAppBar>
      </Box>
    </>
  );
}
