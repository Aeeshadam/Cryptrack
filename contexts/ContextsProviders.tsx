import ClientProvider from "./StoreProvider";
import { AppRouterCacheProvider } from "@mui/material-nextjs/v13-appRouter";
import { ThemeProvider } from "@mui/material/styles";
import theme from "@/styles/theme";
import CssBaseline from "@mui/material/CssBaseline";
import { SearchProvider } from "./SearchContext";
import { InitialDataProvider } from "./InitialDataContext";

export default function ContextsProviders({
  children,
}: Readonly<{
  children: React.ReactNode;
}>) {
  return (
    <html lang="en">
      <body>
        <AppRouterCacheProvider>
          <ThemeProvider theme={theme}>
            <CssBaseline />
            <ClientProvider>
              <SearchProvider>
                <InitialDataProvider>{children}</InitialDataProvider>
              </SearchProvider>
            </ClientProvider>
          </ThemeProvider>
        </AppRouterCacheProvider>
      </body>
    </html>
  );
}
