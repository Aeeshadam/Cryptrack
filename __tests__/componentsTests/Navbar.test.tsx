import React from "react";
import { render, screen, fireEvent } from "@testing-library/react";
import Navbar from "../../components/Navbar";
import { useAuth } from "@/contexts/AuthContext";
import { useRouter } from "next/navigation";
import { useSearch } from "@/contexts/SearchContext";

jest.mock("next/navigation", () => ({
  useRouter: jest.fn(),
}));

jest.mock("@/contexts/AuthContext", () => ({
  useAuth: jest.fn(),
}));

jest.mock("@/contexts/SearchContext", () => ({
  useSearch: jest.fn(),
}));

describe("Navbar", () => {
  const mockPush = jest.fn();

  beforeEach(() => {
    (useRouter as jest.Mock).mockReturnValue({
      push: mockPush,
    });

    (useSearch as jest.Mock).mockReturnValue({
      search: "",
      setSearch: jest.fn(),
    });
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it("should render correctly", () => {
    (useAuth as jest.Mock).mockReturnValue({ user: null });
    const { container } = render(<Navbar />);
    expect(container).toMatchSnapshot();
  });

  it("should render correctly when user is logged in", () => {
    (useAuth as jest.Mock).mockReturnValue({
      user: { uid: "123", email: "demo@gmail.com" },
      displayName: "Demo Demo",
    });
    render(<Navbar />);
    expect(screen.getByText("DD")).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "Portfolio" })
    ).toBeInTheDocument();
  });

  it("should navigate to portfolio page when Portfolio button is clicked", () => {
    (useAuth as jest.Mock).mockReturnValue({
      user: { uid: "123", email: "demo@gmail.com" },
    });
    render(<Navbar />);
    fireEvent.click(screen.getByRole("button", { name: "Portfolio" }));
    expect(mockPush).toHaveBeenCalledWith("/portfolio");
  });

  it("should navigate to sign-in page when Get Started button is clicked", () => {
    (useAuth as jest.Mock).mockReturnValue({ user: null });
    render(<Navbar />);
    fireEvent.click(screen.getByRole("button", { name: "Get Started" }));
    expect(mockPush).toHaveBeenCalledWith("/sign-in");
  });
});
