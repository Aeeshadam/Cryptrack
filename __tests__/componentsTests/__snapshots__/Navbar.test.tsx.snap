// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar should render correctly 1`] = `
<div>
  <div
    class="MuiBox-root css-i9gxme"
  >
    <header
      class="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionStatic css-1q241bj-MuiPaper-root-MuiAppBar-root"
    >
      <div
        class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-188m7n6-MuiToolbar-root"
        style="padding: 0px;"
      >
        <a
          href="/"
        >
          <img
            alt="Logo"
            data-nimg="1"
            decoding="async"
            fetchpriority="high"
            height="50"
            src="/Logo.svg"
            style="color: transparent;"
            width="150"
          />
        </a>
        <div
          class="MuiBox-root css-i9gxme"
        />
        <div
          class="MuiBox-root css-otkr09"
        >
          <div
            class="css-hnq50b"
          >
            <div
              class="css-1u2w42g"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </div>
            <div
              class="MuiInputBase-root MuiInputBase-colorPrimary Mui-readOnly MuiInputBase-readOnly css-621rcj-MuiInputBase-root"
            >
              <input
                aria-label="search"
                class="MuiInputBase-input Mui-readOnly MuiInputBase-readOnly css-yz9k0d-MuiInputBase-input"
                placeholder="Searchâ€¦"
                readonly=""
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
        <div
          class="MuiBox-root css-1kbdbw3"
        >
          <div
            class="css-4bygs5"
            data-testid="mobile-search"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="SearchIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
              />
            </svg>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeSmall MuiButton-textSizeSmall MuiButton-colorPrimary css-1ia8w74-MuiButtonBase-root-MuiButton-root"
          tabindex="0"
          type="button"
        >
          Get Started
          <span
            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
          />
        </button>
      </div>
    </header>
  </div>
</div>
`;
